coreneuron/io/core2nrn_data_return.cpp

```diff
@@ -0,0 +1,174 @@
+ *  of pointers in mdata. tid is the thread index.
+ */
+size_t (*nrn2core_type_return_)(int type, int tid,
+  double*& data, double**& mdata);
+
+namespace coreneuron {
+
+/** @brief permuted array copied to unpermuted array
+ *  If permute is NULL then just a copy
+ */
+static void inverse_permute_copy(size_t n, double* permuted_src, double* dest,
+                                 int* permute)
+{
+  if (permute) {
+    for (size_t i = 0; i < n; ++i) {
+      dest[i] = permuted_src[permute[i]];
+    }
+  }else{
+    std::copy(permuted_src, permuted_src + n, dest);
+  }
+}
+
+/** @brief SoA permuted mechanism data copied to unpermuted AoS data.
+ *  dest is an array of n pointers to the beginning of each sz length array.
+ *  src is a contiguous array of sz segments of size stride. The stride
+ *  may be slightly greater than n for purposes of alignment.
+ *  Each of the sz segments of src are permuted.
+ */
+static void soa2aos_inverse_permute_copy(size_t n, int sz, int stride,
+  double* src, double** dest, int* permute)
+{
+  // src is soa and permuted. dest is n pointers to sz doubles (aos).
+  for (size_t instance = 0; instance < n; ++instance) {
+    double* d = dest[instance];
+    double* s = src + permute[instance];
+    for (int i = 0; i < sz; ++i) {
+      d[i] = s[i*stride];
+    }
+  }
+}
+
+/** @brief SoA unpermuted mechanism data copied to unpermuted AoS data.
+ *  dest is an array of n pointers to the beginning of each sz length array.
+ *  src is a contiguous array of sz segments of size stride. The stride
+ *  may be slightly greater than n for purposes of alignment.
+ *  Each of the sz segments of src have the same order as the n pointers
+ *  of dest.
+ */
+static void soa2aos_unpermuted_copy(size_t n, int sz, int stride,
+  double* src, double** dest)
+{
+  // src is soa and permuted. dest is n pointers to sz doubles (aos).
+  for (size_t instance = 0; instance < n; ++instance) {
+    double* d = dest[instance];
+    double* s = src + instance;
+    for (int i = 0; i < sz; ++i) {
+      d[i] = s[i*stride];
+    }
+  }
+}
+
+/** @brief AoS mechanism data copied to AoS data.
+ *  dest is an array of n pointers to the beginning of each sz length array.
+ *  src is a contiguous array of n segments of size sz.
+ */
+static void aos2aos_copy(size_t n, int sz, double* src, double** dest) {
+  for (size_t instance = 0; instance < n; ++instance) {
+    double* d = dest[instance];
+    double* s = src + (instance*sz);
+    std::copy(s, s + sz, d);
+  }
+}
+      
+/** @brief copy data back to NEURON.
+ *  Copies t, voltage, i_membrane_ if it used, and mechanism param data.
+ */
+void core2nrn_data_return() {
+  if (!nrn2core_type_return_) {
+    return;
+  }
+  for (int tid = 0; tid < nrn_nthread; ++tid) {
+    size_t n = 0;
+    double* data = nullptr;
+    double** mdata = nullptr;
+    NrnThread& nt = nrn_threads[tid];
+
+    n = (*nrn2core_type_return_)(0, tid, data, mdata); // 0 means time
```

Comment:
neuron doesn't have mechanism type `0` and that is the reason we are using first parameter`0` to nrn2core_type_return_?

I wonder if we should change this to some different value (e.g. `< 0`) in case above assumption change?

Reply:
> change this to some different value
Good idea Should add to the enum (on both NEURON and CoreNEURON side).